geomesa-feature/src/main/scala/org/locationtech/geomesa/feature/kryo/KryoSimpleFeatureFactory.scala

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright 2015 Commonwealth Computer Research, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the License);
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an AS IS BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.locationtech.geomesa.feature.kryo
+
+import com.google.common.cache.{CacheBuilder, CacheLoader}
+import org.geotools.factory.{CommonFactoryFinder, Hints}
+import org.geotools.feature.AbstractFeatureFactoryImpl
+import org.geotools.feature.simple.SimpleFeatureBuilder
+import org.geotools.filter.identity.FeatureIdImpl
+import org.locationtech.geomesa.utils.text.{ObjectPoolFactory, ObjectPoolUtils}
+import org.opengis.feature.`type`.AttributeDescriptor
+import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+
+import scala.collection.JavaConversions._
+import scala.ref.SoftReference
+
+class KryoSimpleFeatureFactory extends AbstractFeatureFactoryImpl {
+
+  override def createSimpleFeature(attrs: Array[AnyRef], sft: SimpleFeatureType, id: String) = {
+    val sf = new KryoSimpleFeature(id, sft)
+    sf.setAttributes(attrs)
+    sf
+  }
+
+  override def createSimpleFeautre(attrs: Array[AnyRef], descriptor: AttributeDescriptor, id: String) =
+    createSimpleFeature(attrs, descriptor.asInstanceOf[SimpleFeatureType], id)
+
+}
+
+object KryoSimpleFeatureFactory {
+
+  import scala.collection.mutable.Map
+
+  private val hints = new Hints(Hints.FEATURE_FACTORY, classOf[KryoSimpleFeatureFactory])
+  private val featureFactory = CommonFactoryFinder.getFeatureFactory(hints)
+
+  private val cache = new ThreadLocal[Map[SimpleFeatureType, SoftReference[SimpleFeatureBuilder]]] {
```

Comment:
Nice!  We're getting fancy! 

Lemme see if I have this right...

SimpleFeatureBuilders are not thread safe, hence the ThreadLocal.  We wish to cache them by SFT, so that's the Map.  The SoftReference allows GC to reclaim the SFB, right?  

If that happens, are we reasonably sure that the Map with work appropriately?  (I ask since I haven't used this combination of devices/tricks.)

If all this works as expected, is it (or something close enough) wrapped up into a library?  If this is profitable, part of me wants to see something like
```
class SoftThreadLocalCache[A, B](builder: A => B) { ... }
```

I wonder if one could just ThreadLocal a regular Guava cache?